Implemented lower and upper constraint bounds

Toflar opened the pull request:
Implementation for https://github.com/composer/semver/issues/72 :)

Seldaek commented on tests/Constraint/ConstraintTest.php:
```diff
@@ -158,4 +158,41 @@ public function invalidOperators()
+    public function bounds()
+    {
+        return array(
+            'none' => array('1.0.0.0', '0000001.0000000.0000000.0000000', '1999999.0999999.0999999.0999999'),
```
Lower should be 1.0.0.0, and 1.2.3.4 on the line below 
Upper should be 1.0.0.0 here, 1.2.3.4 below

Seldaek commented on tests/Constraint/ConstraintTest.php:
```diff
@@ -158,4 +158,41 @@ public function invalidOperators()
+            'none' => array('1.0.0.0', '0000001.0000000.0000000.0000000', '1999999.0999999.0999999.0999999'),
+            'none/2' => array('1.2.3.4', '0000001.0000002.0000003.0000004', '1999999.2999999.3999999.4999999'),
+            'parses state' => array('1.0.0.0-RC1-dev', '0000001.0000000.0000000.0000000.RC.0000001.dev', '1999999.0999999.0999999.0999999.9999999.1999999.9999999'),
+            'CI parsing' => array('1.0.0.0-RC15-dev', '0000001.0000000.0000000.0000000.RC.0000015.dev', '1999999.0999999.0999999.0999999.9999999.1599999.9999999'),
```
Lower should be 1.0.0.0-RC15-dev and upper as well as there is no operator

Seldaek commented:
Stopping the review because I notice from the tests it seems quite off from what we'd need to achieve. The operator of the constraint must be taken into account.

Toflar commented:
So what's the actualy goal then here? I thought the goal was to get the normalized bounds from a contstraint so you can `version_compare()` yourself, in which case the constraint is not needed.

Seldaek commented:
Yes that's the goal, but a constraint like  `1.0.0` has upper and lower bounds both `1.0.0.0`, your patch doesn't do this.

While a constraint like `^2.3` has lower bound `2.3.0.0` and upper `2.9999999.9999999.9999999`. Maybe I missed something here, but I don't really see how it allows us to do this?

Toflar commented:
I see. Closing then.

Toflar commented:
Okay, let's check again @Seldaek and @naderman :)

naderman commented on src/Constraint/Constraint.php:
```diff
@@ -212,4 +215,110 @@ public function __toString()
+                $this->bounds[0]['lower'] = implode('.', array_reverse($versionChunks));
+                $this->bounds[0]['upper'] = '9999999.9999999.9999999.9999999.9999999';
+                break;
+            case self::OP_NE:
```
I think you're still thinking about this in a too complex way. We don't need multiple bounds. The lower bound for OP_NE is 0 and the upper bound is 9999999.9999999.9999999.9999999.9999999, that's all. A constraint always has only one lower and one upper bound, there cannot be multiple.

Toflar replied:
Same as above, I want multiple bounds here because that's the correct way to go :) You can still ignore them for your `autoload.php` check case but the library itself should extract all the information so it can also be used for other use cases imho.

naderman replied:
It is not correct. The whole point of boundaries is to be boundaries. If there are values allowed which are higher than the upper boundary, then it is not an upper boundary.

You're really trying to implement something different from what this ticket is about. It's about finding a lower and an upper boundary, not some kind of interval overlap simplifcation, just boundaries.

If you don't want to implement the ticket, I suggest you make a new one for whatever it is you want to build and make a PR for that, and we just keep the original ticket open, because this is not what this ticket was about.

naderman commented on src/Constraint/BoundsProvidingInterface.php:
```diff
@@ -0,0 +1,31 @@
+interface BoundsProvidingInterface
+{
+    /**
+     * Returns an array of bounds.
```
Should return 2 integers, nothing more, a lower bound and an upper bound.

Toflar replied:
No, that's not possible for MultiConstraint and != operators. Your goal is still perfectly achievable with this variant. I don't want to build a "good enough" solution here. We can still do that when writing the bounds into the autoload.php.

naderman replied:
Every constraint has exactly one upper and one lower boundary. That's what the meaning of the word boundary is. It's perfectly possible to do this, if you didn't try to calculate something other than boundaries in this PR.

If you don't want to build a tool for calculating upper/lower boundaries, I suggest you rename this and make a separate PR.

naderman commented on src/Constraint/BoundsProvidingInterface.php:
```diff
@@ -0,0 +1,31 @@
+
+namespace Composer\Semver\Constraint;
+
+interface BoundsProvidingInterface
```
I'm not really sure what the purpose of this interface is?

Toflar replied:
To make sure the constraint provides the new method without breaking BC. We'll also need to implement this on the `MultiConstraint`.

naderman replied:
I'm still unclear why we would need this? Shouldn't this just be a part of the constraint interface?

Seldaek replied:
The point is BC purity as adding a method to the interface breaks implementors, but I'd say we can bump to 2.0 instead.

Toflar replied:
I'm fine with that, I only added it because I thought we want to stay BC. Will remove the interface then.

naderman commented on tests/Constraint/ConstraintTest.php:
```diff
@@ -158,4 +158,60 @@ public function invalidOperators()
+        return array(
+            'equal to 1.0.0.0' => array('==', '1.0.0.0', array(array('lower' => '1.0.0.0', 'upper' => '1.0.0.0'))),
+
+            'lower than 0.0.4.0' => array('<', '0.0.4.0', array(array('lower' => '0', 'upper' => '0.0.3.9999999'))),
```
I think it may actually be easier to just define wether the bound is inclusive or exclusive, than to calculate the lower number.

So e.g. `< 0.0.4.0` would resolve to `['lower' => [INCLUSIVE, '0'], 'upper' => [EXCLUSIVE, '0.0.4.0']]` while `<= 0.0.4.0` would resolve to `['lower' => [INCLUSIVE, '0'], 'upper' => [INCLUSIVE, '0.0.4.0']]`, with some INCLUSIVE/EXCLUSIVE constants defined as true/false I suppose. Rather than using 0 and all nines as min/max you could also literally use INT_MIN/INT_MAX or some constant representing these, then one can simply skip checking the lower/upper bound entirely if that is the value.

Toflar replied:
Okay, given that fact that you only want to use it for `version_compare()` directly, we might as well just provide the value and the operator.
So given we had `< 7.4`, `$constraint->getLowerBound()` would return `['>=', '0']` and `$constraint->getUpperBound()` would return `['<', '7.4']`. Then you can use the result to generate your `version_compare()` statements directly without even handling some `INCLUSIVE` or `EXCLUSIVE` constants?
The only thing that remains an issue is the `9999999...........9999999` upper bound which I would just replace by an infinity constant so one can skip those because they will always be true.
I don't think we need one for the lower boundary because a version lower than 0 just doesn't exist.
Wdyt?

naderman replied:
Just using the operator seems reasonable too! I agree with you on infinity/0 too!

Toflar replied:
So it should be as simple as the latest 1aa9738 now.

Toflar commented:
In case we agree that the current state is the correct way to go, I'd finish this by implementing it on the `MultiConstraint` with some integration tests together with the `VersionParser`.

Seldaek commented:
I think so :)

Toflar commented:
Here we go, looks pretty good imho :) Ping @Seldaek @naderman

Seldaek commented on src/Constraint/MultiConstraint.php:
```diff
@@ -117,4 +123,84 @@ public function __toString()
+                break;
+            case '<':
+                $version .= '.3';
+                break;
```
Can't tell if this function is smart or an atrocity :p Is this using 3/4/6/7 as a sort mechanism of some kind to make version_compare return what you want in case the two bounds are the same? I am not sure if I got it correctly or not, but this definitely needs some comments because it looks super weird.

Toflar replied:
Pff, I did it so it must be smart 😄 I'm kidding - of course this is some black magic here and I don't really like it either. The problem is that we have to compare not only the version but also the operator. Because it does matter whether it's `>=` or `>`. That's a problem now because we went from calculating the actual bounds (+-1) in the `Constraint` to that `[operator, version]` format.
By adding another version for the operator I can just use `version_compare()` internally instead of handling it manually. But I agree that this is hacky. I mean it works but I have to find a nicer solution here. I actually do have an idea. Let me see if I can make this work.

naderman commented on src/Constraint/MultiConstraint.php:
```diff
@@ -117,4 +123,84 @@ public function __toString()
+
+        foreach ($this->constraints as $constraint) {
+            if (!$constraint instanceof BoundsProvidingInterface) {
+                continue; // TODO: or exception? Is it better to ignore those or not?
```
If we just make this part of all constraints there is no need for this?

Toflar replied:
Yes, if we bump to v2 there's no need for that.

naderman replied:
Do we need to go to v2 to add this method to the interface? Do we really guarantee that external implementations of this interface remain compatible?

Toflar replied:
Well, adding a method to the interface is a BC break. I'm not sure what your policy here really is but this library is used by many (https://packagist.org/packages/composer/semver/dependents?order_by=downloads). I have used it standalone as well (never added my own constraint though).

naderman commented on src/Constraint/MultiConstraint.php:
```diff
@@ -117,4 +123,84 @@ public function __toString()
+                continue;
+            }
+
+            if ($this->versionCompare($constraintLower, $this->lowerBound, $this->isConjunctive() ? '>' : '<')) {
```
I think it'd be nicer to actually distinguish the comparison operator here and just do a regular version compare on the versions, rather than the weird version compare / prepare bound magic below?

Toflar replied:
Yeah, let's discuss that below :)

Seldaek commented:
I pushed a test in the branch (sorry broke 5.3 don't wanna push again if you're back working on it), and added another test locally but it fails so I thought I'll let you add it and try to make it pass:

```php
    public function testMultipleMultiConstraintsMergingWithGaps()
    {
        $versionParser = new VersionParser();
        $constraints = [
            '^7.1.15 || ^7.2.3',
            '^7.2.2',
        ];
        foreach ($constraints as &$constraint) {
            $constraint = $versionParser->parseConstraints($constraint);
        }

        $constraint = new MultiConstraint($constraints);

        $this->assertSame(array('>=', '7.2.3.0.dev'), $constraint->getLowerBound(), 'Expected lower bound does not match');
        $this->assertSame(array('<', '8.0.0.0.dev'), $constraint->getUpperBound(), 'Expected upper bound does not match');
    }
```

The lower bound returned is `>=7.1.15.0.dev` which is correct for the first constraint but not for the set of constraints.

Toflar commented:
No prob, will add that test as well.

Toflar commented:
@Seldaek your test is just wrong. The reference thing doesn't work because it would nest constraints and also the lower bound in that case is `7.2.2.0.dev` not `7.2.3.0.dev` 😄 
Adjusted in my latest try though.

Toflar commented:
So I've removed the black magic, the additional interface is gone as well and we now have a proper `Bound` object that represents a bound and also does the comparison to another bound.
Now I am happy :smi

Toflar commented:
I also got rid of the infinity thing and used `PHP_INT_MAX` because otherwise everybody would need to handle the infinity thing and because there are no Composer libraries loaded in our `autload.php` case it would mean to compare to `inf` which is kind of nonsense.

Seldaek commented:
You're right it should have been `>=7.2.2.0.dev` as lower bound, but I don't get what you mean by "The reference thing doesn't work because it would nest constraints". Anyway I see the new version returns 7.2.2 so I guess you fixed it?

Toflar commented:
I fixed your test, yes :)

Seldaek commented:
Well with the previous version I got this:

```
Failed asserting that Array &0 (
    0 => '>='
    1 => '7.1.15.0.dev'
) is identical to Array &0 (
    0 => '>='
    1 => '7.2.3.0.dev'
).
```

So clearly you fixed more than the test..

Toflar commented:
No, that was because of your test 😄 Because you used references and the variables were named the same, the last constraint was a nested constraint again:

```php
    public function testMultipleMultiConstraintsMergingWithGaps()
    {
        $versionParser = new VersionParser();
        $constraints = [
            '^7.1.15 || ^7.2.3',
            '^7.2.2',
        ];
        foreach ($constraints as &$constraint) {
            $constraint = $versionParser->parseConstraints($constraint);
        }

        // Here $constraint is still a reference to the last $constraint of the loop so it's nested again
        // I could have fixed it by doing this here:
        // $multiConstraint = new MultiConstraint($constraints);
        // but I went with the easier solution ;-)
        $constraint = new MultiConstraint($constraints);
    }

Seldaek commented:
Ah that's what you were talking about with references, sorry totally forgot I put a `&` there so I didn't understand anything. Makes sense now! 

Anyway.. you wanna work on #71 based on this? ;) I guess something dumping a set of version_compare statements (+ possibly some normalization code for the runtime version) for the bounds of a given constraint?

Toflar commented:
Guess that's its own PR then.

naderman commented:
@Seldaek #71 is meant as entirely separate thing that actually does precise matching, not meant to be used in combination with this PR here. @Toflar you could add a code generator for checking the bounds in this PR here I think.

Toflar commented:
You mean something that outputs `version_compare(.....) && version_compare()`)?

naderman commented:
@Toflar exactly :-)

naderman commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * @var bool
+     */
+    private $isInclusive;
```
Why did you switch to this now? Thought the operator was pretty good, no?

Toflar replied:
Bounds an always only represent `>`, `>=` or `<` and `<=`. I wanted to enforce this in the library. The array solution allowed random values and I like being explicit about things so I went for a proper class representing exactly what's possible only.

naderman replied:
Sure I wasn't questioning the class anyway, just whether a bool vs operator made sense, but it's fine by me :-)

naderman commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        // Not the same version means we can use regular version comparison and inclusive/exclusive doesn't matter
+        if (!version_compare($this->getVersion(), $other->getVersion(), '==')) {
```
Could avoid running version compare twice here, by just running it once without an operator and using the result variable?

Toflar replied:
I need to know two results. 1. if they are equal and 2. if the match the given operator. So it will always be two operators = two comparisons.

naderman replied:
No need, you can just do version_compare and then check if negative, positive or 0 and compare to the operator?

Toflar replied:
Ah, I misunderstood this one. You're right. Fixed in 83cc74a7de4008993754ebb591fc6cb5f1d7344d.

GrahamCampbell commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,96 @@
+<?php
+
+
```
Missing header comment.

GrahamCampbell commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,96 @@
+
+namespace Composer\Semver\Constraint;
+
+
```
Too many new lines.

GrahamCampbell commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,96 @@
+    {
+        return new Bound((string) PHP_INT_MAX, false);
+    }
+}
```
Missing newline.

GrahamCampbell commented on tests/Constraint/ConstraintTest.php:
```diff
@@ -188,4 +188,62 @@ public function invalidOperators()
+            'not equal to 1.0.0.0' => array('<>', '1.0.0.0', Bound::lowerMost(), Bound::upperMost()),
+        );
+    }
+}
```
Missing newline.

Seldaek commented:
@GrahamCampbell we have php-cs-fixer for this kind of stuff, so maintainers can fix things whenever. Please refrain from spamming CS comments.

Toflar commented:
I was just gonna say, can we use the CS fixer on master and implement it in CI for that? I don't know the preferences of this project so...

GrahamCampbell commented:
I don't see any code style checking on the CI, otherwise I'd have not commented.

Seldaek commented:
I am against CI checks for code style, it's annoying. I run fixer every now and then to normalize things. No big deal.

Toflar commented:
I would've let it run on every commit. I'm so used to it ;) Feel free to add your perferred config on master, I'll happily rebase then.

Seldaek commented:
You can run `php-cs-fixer fix` if you want to fix things here.

Toflar commented:
Ah, I didn't notice it was there all the time because the cs fixer is not a dev requirement ^^
Will do next time I push.

naderman commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,106 @@
+
+        // Not the same version means we don't need to check if the bounds are inclusive or not
+        if (0 !== $compareResult) {
+            return ('>' === $operator ? 1 : -1) === $compareResult;
```
Please add another set of brackets around the first comparison, so one doesn't need to remember ternary operator precdence:

```php
return (('>' === $operator ? 1 : -1) === $compareResult;
```

Toflar replied:
Done.

Toflar commented on src/Constraint/Constraint.php:
```diff
@@ -212,4 +220,96 @@ public function __toString()
+     *
+     * @param string $version
+     */
+    public static function getVersionCompareCondition(ConstraintInterface $constraint, $version)
```
@Seldaek @naderman Doesn't really belong here and should go to its own class maybe. But how do we name it? And is the `VersionParser` injected or do we expect the `$version` to be normalized when passing?
It needs to be normalized because `version_compare('1.0', '1.0.0.0', '>=')` is `false`:(

naderman replied:
Hm that's a bit of a problem, we don't want to be normalizing PHP versions in our autoloader use case?

Toflar replied:
Jup but I cannot change the way PHP works in `version_compare()` 😄

Toflar replied:
And yes, the returned string should have `%s` and so it can inject the version at runtime and obviously not pre-generate it like it is now. Lol. But still, the main issue remains.

Seldaek replied:
I thought this was the point of the stuff at https://github.com/composer/semver/pull/75/files#diff-d094db696fbeed03e5e71f05bdb50ceeR282-R286 that it can be replicated easily at runtime without needing full normalizer.. Then could just pad with .0 until we have 4 dots?

Toflar replied:
Normalizing just does what PHP does internally in `version_compare()`. See https://github.com/php/php-src/blob/6f9248359e5a0f9056e198a7ad9f87ca6e78aa56/ext/standard/versioning.c#L44-L46

Seldaek replied:
As for a class, I don't know call it Composer\Semver\ComparisonDumper or smth? :) 

And would be good to only dump a single version_compare if one of the bounds is infinite

Seldaek replied:
What about the:

> Then could just pad with .0 until we have 4 dots?

Toflar replied:
The problem is not that semver needs to pad. The one comparing to it would need to pad.
You would need to normalize `PHP_VERSION` at runtime.

Toflar replied:
In fact I could remove that part now that we're not calculating bounds ourselves anymore.

Seldaek replied:
I realize that, I'm saying the generated code should pad the $version/%s so that it matches the normalized/padded constraints. Kinda do a best effort minimal version of the normalization that hopefully is enough to make version_compare work.

Toflar replied:
I got rid of the canonicalization stuff and implemented a ComparisonDumper which is also optimized to not dump lower and upper most bounds (so if both are given, it just dumps `true`).

naderman commented on src/Constraint/ComparisonDumper.php:
```diff
@@ -0,0 +1,78 @@
+     * whether a version matches upper and lower bounds.
+     *
+     * @param ConstraintInterface $constraint
+     * @param string              $version
```
I don't understand why you are passing a constraint and a version? Shouldn't you only be passing a constraint to this?

Toflar replied:
Depends. In case of the planned PHP version check we have access to this library and can generate the complete string right away. In case you only want to dump is and add the version later, we‘d need `%s` so you can sprintf(), I guess.

Seldaek replied:
We don't have the runtime version when composer runs as that might be different php/env. That's the whole point of the feature.

Toflar replied:
So just a placeholder then, right?

Toflar replied:
Okay, implemented in 5bb8bc8. We do have to normalize the versions at runtime which is why the generated code needs quite a bit of logic but I really kept it down to the bare minimum:

* Only the runtime version is normalized at runtime, the bounds are when dumping the comparison.
* Only what's really needed is dumped, no variables for upper or lower bounds if they are lowermost or uppermost.

Using `eval()` here. We could also expect to just set `$version` before the dump, so we don't have to use `eval()`.

Seldaek replied:
eval is fine in the test IMO. It should end up being dumped in a file that gets executed as usual anyway. What I'd like tho is that %version% would not be wrapped in quotes in the dumped code, that way it's easy to replace it with smth like PHP_VERSION or whatever is needed.

Toflar replied:
Right, done in 108aad5d81300af55c305a89d4ece459beb52c44.

Seldaek commented:
Yes I think if you know the version there is not much point in dumping the
code you rather use the constraint class directly to check and move on.

Toflar commented:
It was a long day...

Seldaek commented:
I'm not saying you have to do this now :) a beer and sleep would be more appropriate at this point.

Toflar commented:
After some discussions with @Seldaek and @naderman, we agreed that the whole comparison logic is probably way too specific and out of scope for the semver component.
I've updated the PR accordingly and rebased it onto master. It now contains only the bounds logic, so this should be ready for a final review. Failing tests seem unrelated (probably just needs a restart). Want me to squash or will you be squash merging anyway?

naderman commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,116 @@
+        return $this->isInclusive;
+    }
+
+    public function isLowerMost()
```
rename to "isLowest"?

naderman replied:
or maybe "isMin"?

naderman replied:
rename to "isZero"

naderman commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,116 @@
+        return $this->getVersion() === '0' && $this->isInclusive();
+    }
+
+    public function isUpperMost()
```
rename to "isHighest"?

naderman replied:
or if you go for isMin, isMax?

naderman replied:
rename to "isPositiveInfinity()"

naderman commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * @return self
+     */
+    public static function upperMost()
```
Should this be called positiveInfinity? or something? or max? "upperMost" is pretty weird.

naderman replied:
rename to "positiveInfinity()"

naderman commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * @return self
+     */
+    public static function lowerMost()
```
rename to "zero()"

stof commented on src/Constraint/Bound.php:
```diff
@@ -0,0 +1,116 @@
+        }
+
+        // Question we're answering here is "am I higher than $other?"
+        return '>' === $operator ? $other->isInclusive() : !$other->isInclusive();
```
this looks weird to be that `$this->isInclusive()` is never taken into account. Is this correct ?

Toflar replied:
tbh I don't know. I've been messing with this code for too long but there are lots of tests in the `MultiConstraintTest` which uses the `compareTo()` method internally. So maybe we were missing a constellation that requires a test? Would be great if you could provide that. More eyes is always better :)

naderman replied:
`$this->isInclusive` is implicitly checked when we do `($this == $other)` above.

When we come to this part of the code, the version on both bounds is the same, so the following table is the result (x means, the bounds are identical, so the early return at the top of the function prevents this from mattering):

|expression|this.isInclusive|other.isInclusive|result|
|-|-|-|-|
|this.compareTo(other, >)|0|0|x|
|this.compareTo(other, >)|0|*1*|*1*|
|this.compareTo(other, >)|1|*0*|*0*|
|this.compareTo(other, >)|1|1|x|
|this.compareTo(other, <)|0|0|x|
|this.compareTo(other, <)|0|*1*|*0*|
|this.compareTo(other, <)|1|*0*|*1*|
|this.compareTo(other, <)|1|1|x|

So you can see that the expression is correct. But since we exclude a large part of the variations with early returns, an expression like `'>'  === $operator ? !$this->isInclusive() : $this->isInclusive()` would be just as correct, since we already made sure that `$this->isInclusive() != $other->isInclusive()` once it gets to this section of the code.
